trace/trace.go

```diff
@@ -157,11 +171,156 @@ func decodeHex(h string, b []byte) error {
+		sb.WriteString((string)(kv.Key))
+		sb.WriteByte('=')
+		sb.WriteString(kv.Value.Emit())
+
+		if i != len(ts.kvs)-1 {
+			sb.WriteByte(',')
+		}
+	}
+
+	return sb.String()
+}
+
+// Get returns a value for given key from the trace state.
+// If no key is found or provided key is invalid, returns an empty value.
+func (ts *TraceState) Get(key label.Key) label.Value {
+	if ts == nil || !isTraceStateKeyValid(key) {
+		return label.Value{}
+	}
+
+	for _, kv := range ts.kvs {
+		if kv.Key == key {
+			return kv.Value
+		}
+	}
+
+	return label.Value{}
+}
+
+// Insert adds a new key/value, if one doesn't exists; otherwise updates the existing entry.
+// The new or updated entry is always inserted at the beginning of the TraceState, i.e.
+// on the left side, as per the W3C Trace Context specification requirement.
+func (ts *TraceState) Insert(entry label.KeyValue) (TraceState, error) {
+	if !isTraceStateKeyValueValid(entry) {
+		return TraceState{}, errInvalidTraceStateKeyValue
+	}
+
+	if ts == nil {
+		return TraceState{[]label.KeyValue{entry}}, nil
+	}
+
+	ckvs := ts.copyKVsAndDeleteEntry(entry.Key)
+	if len(ckvs)+1 > traceStateMaxListMembers {
+		return TraceState{}, errInvalidTraceStateMembersNumber
+	}
+
+	ckvs = append(ckvs, label.KeyValue{})
+	copy(ckvs[1:], ckvs)
+	ckvs[0] = entry
+
+	return TraceState{ckvs}, nil
+}
+
+// Delete removes specified entry from the trace state.
+func (ts *TraceState) Delete(key label.Key) (TraceState, error) {
+	if ts == nil {
+		return TraceState{}, nil
+	}
+
+	if !isTraceStateKeyValid(key) {
+		return TraceState{}, errInvalidTraceStateKeyValue
+	}
+
+	return TraceState{ts.copyKVsAndDeleteEntry(key)}, nil
+}
+
+func (ts *TraceState) copyKVsAndDeleteEntry(key label.Key) []label.KeyValue {
+	ckvs := make([]label.KeyValue, len(ts.kvs))
+	copy(ckvs, ts.kvs)
+	for i, kv := range ts.kvs {
+		if kv.Key == key {
+			ckvs = append(ckvs[:i], ckvs[i+1:]...)
+			break
+		}
+	}
+
+	return ckvs
+}
+
+// TraceStateFromKeyValues is a convenience method to create a new TraceState from
+// provided key/value pairs.
+func TraceStateFromKeyValues(kvs ...label.KeyValue) (*TraceState, error) { //nolint:golint
+	if len(kvs) > traceStateMaxListMembers {
+		return nil, errInvalidTraceStateMembersNumber
+	}
+
+	km := make(map[label.Key]bool)
+	for _, kv := range kvs {
+		if !isTraceStateKeyValueValid(kv) {
+			return nil, errInvalidTraceStateKeyValue
+		}
+		_, ok := km[kv.Key]
+		if ok {
+			return nil, errInvalidTraceStateDuplicate
+		}
+		km[kv.Key] = true
+	}
+
+	return &TraceState{kvs}, nil
+}
+
+func isTraceStateKeyValid(key label.Key) bool {
+	return keyFormatRegExp.MatchString(string(key))
+}
+
+func isTraceStateKeyValueValid(kv label.KeyValue) bool {
+	return isTraceStateKeyValid(kv.Key) &&
+		valueFormatRegExp.MatchString(kv.Value.Emit())
+}
+
 // SpanContext contains identifying trace information about a Span.
 type SpanContext struct {
 	TraceID    TraceID
 	SpanID     SpanID
 	TraceFlags byte
+	TraceState *TraceState
```

Comment:
Can this be a struct, not a pointer? It should simplify the code a bit.

Reply:
I think it could be. It would require some extra work, as `SpanContext` would no longer be comparable (as at some test utils we use it as [map key](https://github.com/open-telemetry/opentelemetry-go/blob/master/oteltest/span.go#L50) or in SDK [in comparison](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/trace/span.go#L316)), but seems like its doable just fine.

Reply:
The use of a pointer in this case means that using a SpanContext with a map key can never be equal to another. Do users want a comparable SpanContext like this? Is this use of SpanContext as a map key for more than testing purposes? 

In my feedback elsewhere in this PR, it's occurred to me that some of the `label/set.go` code could be used to implement trace state. At that point we could offer users a form of span context that works as a map key and supports order-independent trace state equality. I won't say this is possible w/o modifications/extensions in the label.Set code, but it's appealing to me nevertheless (e.g., the `label.Set` method supports iteration but is not directly usable as a map key, `label.Distinct` works as a map key but does not support iteration, and there's no support for deletion as a primitive).